pkg/scheduler/plugins/noderesourcefitplus/node_resources_fit_plus.go

```diff
@@ -0,0 +1,202 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package noderesourcesfitplus
+
+import (
+	"context"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/kubernetes/pkg/api/v1/resource"
+	k8sConfig "k8s.io/kubernetes/pkg/scheduler/apis/config"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+	plfeature "k8s.io/kubernetes/pkg/scheduler/framework/plugins/feature"
+	"k8s.io/kubernetes/pkg/scheduler/framework/plugins/noderesources"
+
+	"github.com/koordinator-sh/koordinator/pkg/scheduler/apis/config"
+)
+
+const (
+	// Name is plugin name
+	Name = "NodeResourcesFitPlus"
+
+	preScoreStateKey = "PreScore" + Name
+)
+
+var (
+	_ framework.ScorePlugin = &Plugin{}
+)
+
+type Plugin struct {
+	handle framework.Handle
+	args   *config.NodeResourcesFitPlusArgs
+}
+
+func New(args runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+
+	sampleArgs2, ok := args.(*config.NodeResourcesFitPlusArgs)
+
+	if !ok {
+		return nil, fmt.Errorf("want args to be of type NodeResourcesArgs, got %T", args)
+	}
+
+	return &Plugin{
+		handle: handle,
+		args:   sampleArgs2,
+	}, nil
+}
+
+func (s *Plugin) Name() string {
+	return Name
+}
+
+func (s *Plugin) PreScore(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodes []*v1.Node) *framework.Status {
+	cycleState.Write(preScoreStateKey, computePodResourceRequest(pod))
+	return nil
+}
+
+func (s *Plugin) Score(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) (int64, *framework.Status) {
+
+	nodeInfo, err := s.handle.SnapshotSharedLister().NodeInfos().Get(nodeName)
+	if err != nil {
+		return 0, framework.NewStatus(framework.Error, fmt.Sprintf("getting node %q from Snapshot: %v", nodeName, err))
+	}
+
+	var nodeScore int64
+	var weightSum int64
+
+	scoreState, err := getPreScoreState(state)
+	if err != nil {
+		return 0, framework.NewStatus(framework.Error, fmt.Sprintf("get State node %q from PreScore: %v", nodeName, err))
+	}
+
+	podRequest, _ := fitsRequest(scoreState.Resource, nodeInfo)
+
+	for _, requestSourceName := range podRequest {
+		v, ok := s.args.Resources[requestSourceName]
+		if !ok {
+			continue
+		}
+		fit, err := noderesources.NewFit(
+			&k8sConfig.NodeResourcesFitArgs{
+				ScoringStrategy: &k8sConfig.ScoringStrategy{
+					Type: v.Type, // MostAllocated or LeastAllocated
+					Resources: []k8sConfig.ResourceSpec{
+						{Name: string(requestSourceName), Weight: 1},
+					},
+				},
+			}, s.handle, plfeature.Features{})
+
+		if err != nil {
+			return 0, framework.NewStatus(framework.Error, err.Error())
+		}
+
+		resourceScore, status := fit.(framework.ScorePlugin).Score(ctx, state, p, nodeName)
```

Comment:
I recommend you just copy the upstream code and update the license "Copyright 2022 The Koordinator Authors" to "Copyright 2022 The Kubernetes Authors"

Reply:
Do you mean to suggest duplicating MostAllocated and LeastAllocated's implementation of Score? Rather than suggesting this form of encapsulating score?

Reply:
> @LY-today It seems to be a performance issue due to the frequent creation of the noderesourcefit plugin in the Score phase.

It seems that we can't use encapsulation, but copy the original code and re-implement it?

Reply:
done